### ISSUE TITLE
works well on smaller models, declines as model size increases and not competitive at 8B

### ISSUE THREAD

**Issue by lessw2020** [zyushun/Adam-mini#7]
Hi there,
First, really nice job on developing adam-mini!  It's a really refreshing approach for adam.
I did some initial testing by integrating adam_mini into torchtitan, and ran it with varying sizes of hand-made llama3 models from 10M to 8B (official size). 

What I found is that adam-mini is very competive if not better than AdamW at small model sizes, but declines as the model gets larger.  
At 8B, then it's spiking all over the place during the warmup lr phase, and then kicks in with a more monotonic loss decline, once lr starts it's linear decline...but at that point it is so far behind it cannot compete. 

I'm wondering if there is any thought to further sub-dividing the hessian sub-groups as model size grows, or do you have any other input re: helping adam-mini be effective at say 8B scale?

I did see a large amount of memory savings at 8B (~9.5% total reduction per gpu) so that is promising, but the training accuracy as noted is not competitive at that scale. 

If you are interested in running to experiment, I have posted the code (had to make a couple integration tweaks) here:
https://github.com/lessw2020/torchtitan_oss/tree/adam_mini

**Comment by zyushun** [comment]
Thanks for your interest and thanks for the feedback! We are trying to reproduce your results on 8B now. 

I noticed that you pass the "all_reduce" operation in our original implementation: line 222 of [Adam_mini.py](https://github.com/zyushun/Adam-mini/blob/main/Adam_mini.py). Please try to put "all_reduce" back and see if it helps. Additionally, you can try smaller lr and see if it is better

**Comment by zyushun** [comment]
Hi, 
I just found that our implementation of "all_reduce" is not compatible with your code. We are working on implementing a new "all_reduce" now

**Comment by zyushun** [comment]
Hi, further update:

I just found that you are actually using  (1) **default pytorch partition (2) empty embd_blocks (3) without all_reduce**.  All these will cause training instability. Why does (1) and (2) happen? this is because: we currently judge  "whether the current block is embedding or Q or K " by the name of that block (i.e., "attn.wq.weight"). The name in your code is different from ours (i.e., you are using  "atttention.wq.weight" instead of "attn.wq.weight" ), so all our if-else condition fails. Sorry for the inconvenience. Will update together with the compatible  all_reduce operator.

**Comment by lessw2020** [comment]
Hi @zyushun!
Thanks very much for investigating this and very glad to hear that you've isolated things re: titan and llama3.
Will be happy to run things if you have an update both for 8B and then 70B. 

Also, I should add that the reason for the changes in the all_reduce were b/c titan uses DTensor, and so the tmp_lr is a dtensor and thsu cannot be used in a normal all_reduce.  Instead, I had to do force it to do an all_reduce via the line I added.
I did also test a workaround which I think I left in the commments of tmp_lr.to_local() and then regular all_reduce and then back to DTensor but I found the precison was better to simply do the direct all_reduce via DTensor.
I can check on how to avoid that from becoming a DTensor in the first place as well, though I think it works fine as DTensor atm.

Lastly, the main repo for Titan is here (in case you want to make your own fork).  Please ensure you run with latest PyTorch nightly though otherwise the apis are changing a fair amount and older PyTorch builds may break:
https://github.com/pytorch/torchtitan

Please let me know what you find and also if you have further questions about titan/llama3!

**Comment by zyushun** [comment]
Hi @lessw2020!

I just updated a new version of Adam-mini that can work on your codebase. Please use this [latest version](https://github.com/zyushun/Adam-mini/blob/main/Adam_mini.py) it as follows

            optimizer = Adam_mini(model, lr=lr, 
            betas = (0.9,0.95), 
            weight_decay=0.1, 
            model_sharding=True,
            n_embd=4096,
            n_head=32,
            n_query_groups=4
            )

Using your codebase, I pre-trained Llama3-8B with 1000 steps and I got the following curves. Adam-mini performs almost the same as AdamW.  (Remark:  I tuned down the lr of Adam-mini to 1e-4, while the lr of AdamW is remained as 3e-4).

![46461720264536_ pic](https://github.com/zyushun/Adam-mini/assets/85537560/bb1745d5-41f8-4790-891e-708923f5c323)


I use the same setting as you shared expect for the following (minor) changes: 


1. I use seq_len = 2048  since the original seq_len of 8192 caused   OOM error on my (poor) machine. You can change it back on your machine

2. used "c4_mini" as you attached in the code

3. changed "fused_rmsnorm" to "rmsnorm". This is because the  "fused_rmsnorm" is not comptatible with our all_reduce operation for vmean. If you insists on using "fused_rmsnorm", a simple tweak is to use AdamW for  normalization layers as well. This only increases negligible  memory  and I guess it won't affect the performance. You can find these layers by parameter using keyword "norm"  in their name, and put them in the if-else condition for embedding blocks. 


Please do not hesitate to contact us for any further update! If it also works on your side, feel free to try it out on larger models such as 70B! Looking forward to hear from you soon.

Yushun

**Comment by lessw2020** [comment]
Hi @zyushun, this is great news! Let me get things setup and will run and update here.

**Comment by lessw2020** [comment]
Hi @zyushun  - here are some initial results.  It's looking way better with your changes, though AdamW still slightly outperforms. 
Changes:
1 - I used sequence of 4096 (vs you had 2048 as you clarified due to OOM)
2 - I used c4 dataset.  The c4_mini will loop and reuse data for a 1K run on 8 gpus, so we don't recommend c4_mini except for short runs.  C4 is huge so no data looping.
3 - I compared both at 1e-4 to start.  (not shown yet, I also ran with adamw 3e-4 and it performed even better than adamw1e-4, so I have adam_mini running now also with 3e-4...will update soon).
4 - as you noted, I ran with rmsnorm to keep things matching.

Anyway, here's the results for 1e-4 for both:
![Uploading Screenshot 2024-07-06 at 2.48.10 PM.png…]()

**Comment by lessw2020** [comment]
qq - I think I'll rerun next with the same sequence length you had (2048) and see if that is having an impact.  I'm wondering also how num_embeddings in the init for adam_mini is determined?
As you can see, the loss curves are in terms of ups and downs are almost identical with just a slight gap on the y axis. 

Finally memory savings with adam-mini are substantial:
<img width="353" alt="Screenshot 2024-07-06 at 2 59 15 PM" src="https://github.com/zyushun/Adam-mini/assets/46302957/689480ca-0152-4e75-9b16-d4abc660f15f">

**Comment by lessw2020** [comment]
Hi @zyushun  - this looks very promising - comparing adamw 1e-4 with adam-mini 3e-4, their loss curves so far are almost clones:

<img width="1014" alt="Screenshot 2024-07-06 at 3 03 20 PM" src="https://github.com/zyushun/Adam-mini/assets/46302957/2cd490d8-15ab-4301-8c86-dff79433784b">

Will update once the run is finished, but it seems the main thing needed now is just slight adjustments on the lr for mini vs adamw and mini can match adamw.

**Comment by lessw2020** [comment]
one more update - it seems 3e-4 is the optimal value for adamw.  
I tried to match it with mini and 4e-4, as well as 5e-5 with longer warmup (with same warmup of 300 then 5e-5 is unstable) but for these mini ends up ~ .15 behind:

loss curve:
<img width="1139" alt="Screenshot 2024-07-06 at 6 13 20 PM" src="https://github.com/zyushun/Adam-mini/assets/46302957/be997436-cb16-49a8-b584-5974a7a46e9f">

Is there anything else that could be adjusted here - it's very close, but I'm not sure what else might help here.

**Comment by zyushun** [comment]
Hi @lessw2020 ! Thanks so much for the detailed update! We are sincerely grateful for your support and all the great discussion!

I have carefully read all your updates, it seems that: on 8B models:  

 1. when seq_len = 2048: Adamw(3e-4, best tuned)  = Adam-mini (1e-4). Highly similar loss curve.
 
 2. when seq_len = 8192, Adamw(3e-4, best tuned) is a bit faster than Adam-mini(4e-4, best tuned). the gap is about 0.15 in train loss. The shape of the curve still looks highly similar.
 
This is quite interesting. I suspect there is something mysterious occurs in **the long-context scenario**. I will try to get 8 gpus today  and try to reproduce and shrink this 0.15 gap on my side.  I will update to you as soon as I get new findings.

Remark: there is one another possibility: this 0.15 gap will shrink as training goes on. This sometimes happens on some small-scaled task, too. 

Quick answers to some of your other questions:

1. n_embd: this is the hidden feature dimension of your LLM, also known as "hidden_size" in LLama3 series. Note that this is not the vocabulary size (sorry for the potential confusion here). For Llama3-8B, n_embd = 4096; for Llama3-70B n_embd = 8192

2. fused_rmsnorm: We are working on it, but it seems not so trivially solvable. For now, I suggest to try to put all the RMSnorm into "embd_blocks" of Adam-mini. For blocks in  "embd_blocks", AdamW will be used to update parameters so there is no all_reduce operation any more. This can be simply done with one-line of code change:  change [line 134 in Adam-mini.py](https://github.com/zyushun/Adam-mini/blob/main/Adam_mini.py) to the following:

```
 if (
                            "embed_tokens" in name or "wte" in name or "lm_head" in name or "tok_embeddings" in name or "output.weight" in name or "norm" in name):
                        if p.grad is None:
```

This change will only incurr negligible memory increase, so slightly that i belive you will not even notice them. I  haven't try this before but I guess it won't hurt the performance.


 I will try to get 8 gpus today  and reproduce your results on my side. Thanks again for your interests and support. Let us keep in touch.

**Comment by lessw2020** [comment]
Hi @zyushun,
Thanks for the info above!  
I have another update that may help. 

I switched to short context (2084) and reduced the run to 900 (with warmup = 300) so I could run more tests.  I did a sweep of lr from 1e-4 to 5e-4 for mini and 3e-4 with adamW as the control.  
I got different results from you in the respect that the slightly higher lr for mini still performed the best of the lr's for mini.  5e-5 became unstable. 

Here's the overview:
![Uploading Screenshot 2024-07-06 at 9.04.45 PM.png…]()

**Comment by zyushun** [comment]
Hi @lessw2020 ! Thanks for the swift response!

When seq_len = 2048, it seems that my previous result (AdamW3e-4 = Adam-mini1e-4) is a bit different from yours (AdamW3e-4-yellow is slightly better than Adam-mini1e-4).  Is it due to "c4_mini" v.s. "c4"? or different warm_up? I used warm_up = 40 (why 40? tbh I forgot why I chose this number..).  I will try to reproduce your result today.

By comparing the curves for seq_len = 2048 and seq_len = 8192, it seems that Adam-mini is much closer to AdamW in seq_len = 2048 than in seq_len = 8192. Huh, interesting. 

Regarding "n_embd" we adopt this name from Tinyllama. Thanks for the suggestion and we will consider changing it. 

I will keep updating.

**Comment by lessw2020** [comment]
Hi @zyushun  - a bit of progress.
I ran 2400 iters 9800 warmup) and as part of that, the gap for the best point on both curves closed to .1 difference vs closer to .2 in the early iters. 
So I'm kicking off 5000 iter run now and also increasing batch size to 8 since it has room with 2048 seq len. 

Here's the 2400 chart:

<img width="1467" alt="Screenshot 2024-07-06 at 11 01 34 PM" src="https://github.com/zyushun/Adam-mini/assets/46302957/09d9dcc0-1fc4-40f4-8539-d33f7cb8cc46">

Re: your results- I would say the issue there is that using c4_mini means the data has looped and therefore it's less challenging as the data repeats.  So using c4 is a better run from a data integrity aspect. 
Also re: warmup - I'm using much longer warmups (often 25% of the total iters) so maybe that means you spent longer on the downhill of the lr and so results converged better.

Hopefully my 5K run tonight and tomorrow morning will show if that is where the gap can be closed.
Also, thanks for the tip on how to update for rmsnorm.  I will add that in tomorrow but sticking with rmsnorm for now so we have consistent comparisons across all tests today.

**Comment by zyushun** [comment]
Hi @lessw2020 , update on seq_len = 2048

I run 10,000 steps with 2500 warm_up steps as you recommended.  I use lr =  3e-4 for both Adam-mini and AdamW. I still use per-gpu bs = 2. I changed "c4_mini" to "c4" to avoid re-loop. Here is what I got.
 
<img width="664" alt="image" src="https://github.com/zyushun/Adam-mini/assets/85537560/1e672587-45c6-4532-9559-caf468327496">


Adam-mini is slightly slower in the first  2000 steps but then catches up and slightly outperforms AdamW in the remaining steps. As a bonus finding, AdamW suffers from a loss spike around 4000+ step, while Adam-mini does not. But we do not intend to claim any thing about it.

Hope you also got the similar results on your side :D

**Comment by lessw2020** [comment]
Hi @zyushun  - nice, thanks for the 10K data!
I got similar results at 5K in that the early lead by AdamW gets closed and best min loss is only -0.04 diff. 
I ran with larger batch size (8):

<img width="308" alt="Screenshot 2024-07-07 at 1 43 07 PM" src="https://github.com/zyushun/Adam-mini/assets/46302957/76d5ce5f-93d9-419b-a34f-e57ab7542d9e">

And loss curves:
<img width="1484" alt="mini_5K_results" src="https://github.com/zyushun/Adam-mini/assets/46302957/11eec162-6893-42a0-96f7-8643429ca090">

I think I will update mini to use fused_rmsnorm (thanks for the info on how to do this!) and then may do a run with 10K as well and then try mini on 70B tomorrow.

Regardless, it's clear mini is a strong candidate here - I want to see if we can get it added as an optimizer option for titan, and also depending on time, I would be interested to move it into a cuda kernel to further speed it up.  
Let me update it for using fused though and kick off the 10K run as next steps.

**Comment by lessw2020** [comment]
Hi @zyushun  - 
I've updated with your changes to integrate fused rms_norm...I felt the number of checks was confusing though so have simplified it (and should be a tad faster) by using a set:
~~~
# specific layers, including fused_rmsnorm, are incompatible and are run using normal adamW updates
        self.exclude_layers = {
            "embed_tokens", "wte", "lm_head", "tok_embeddings", "output.weight", # "norm"
        }
 ~~~
 and then:
 ~~~
 # handle excluded layer types via adamw updates
                    if any(layer in name for layer in self.exclude_layers):
 ~~~
 
 that is now allowing it to run with fused rmsnorm, which is adding about 4% MFU and a slight further drop in gpu memory...
 Anyway, running with the 10K run now.
 
 One item if you have time - I went to test my fused rmsnorm changes with the debug_model.toml which is our goto for verifying changes before running larger but now I get a shape incompat error.  I verified this is not related to the fused_rmsnorm changes. 
 Could you take a look by switching to running the debug_model.toml in the run_llama_train.sh and see if you are able to understand what is going awry there?

**Comment by zyushun** [comment]
Hi @lessw2020 ! Thanks soooooo much for your great advice! We will modify them immediately.

I am  also working on the debug_model.toml now.

Bad news: as for fused_norm, it seems that my naive solution does not work :(. When putting rmsnorm into "exclude_layers", the loss wont go down.  I will try to work out an alternative solution.

**Comment by zyushun** [comment]
Hi @lessw2020 !  Regarding the debug_model.toml.

The incompat error is because: this debug model uses different n_heads, dim, and so on, while Adam-mini does not know this. This could be fixed as follows.

1. Add  " global model_config" at [line 232 of train.py](https://github.com/lessw2020/torchtitan_oss/blob/adam_mini/train.py).

```
    global model_config
    model_config = models_config[model_name][job_config.model.flavor]
```
2. Change the input of Adam-mini as follows.

```
optimizer = Adam_mini(model, lr=lr, 
            betas = (0.9,0.95), 
            weight_decay=0.1, 
            model_sharding=True,
            n_embd=model_config.dim,
            n_head=model_config.n_heads,
            n_query_groups=model_config.n_kv_heads,
            )
```

Best,
Yushun

**Comment by zyushun** [comment]
Hi @lessw2020 , regarding fused_rmsnorm:

A strange thing happened. I did not change anything but the original Adam-mini seems to work with fused_rmsnorm. See the curve below. Note that I did NOT add "norm" in self.excluded_layers.

<img width="593" alt="image" src="https://github.com/zyushun/Adam-mini/assets/85537560/8e89c056-08f3-4967-bdf9-248be9b8c277">

The log says we are indeed using fused_rmsnorm.

```
[rank0]:2024-07-08 13:40:10,845 - root - INFO - Building llama3 8B with ModelArgs(dim=4096, n_layers=32, n_heads=32, n_kv_heads=8, vocab_size=128256, multiple_of=1024, ffn_dim_multiplier=1.3, norm_eps=1e-05, rope_theta=500000, max_batch_size=32, max_seq_len=2048, depth_init=True, norm_type='fused_rmsnorm')
```

Could you try again on your side to see if it is also working now?

**Comment by lessw2020** [comment]
Hi @zyushun,
Thanks for the updates re: fused norm and the update for making the inputs directly updated.
I was able to run a short experiment tonight and can confirm that fused_rmsnorm is working nicely with mini with no changes.
This is very helpful as fused improves MFU and lowers total memory so it's a good combo!
Did not have time yet to run a 70B run but still planning to do that next.
Best regards,
Less

**Comment by zyushun** [comment]
Hi @lessw2020 .

It is great to hear that things work well on your side, too. Thanks again for all the great discussion and suggestions!  If you encounter any further challenges, please do not hesitate to contact me.

Best,
Yushun

**Comment by zyushun** [comment]
Hi @lessw2020 !

We just update a new version of [Adam_mini.py](https://github.com/zyushun/Adam-mini). This version integrated your suggestions  and some other minor refinements. Please remember to update this [Adam_mini.py](https://github.com/zyushun/Adam-mini) in your local codebase.  You can call as follows:
```
from Adam_mini import Adam_mini

optimizer = Adam_mini(
		model = model, 
		lr = lr, 
		betas = (beta1,beta2), 
		eps = eps,
		weight_decay = weight_decay,
		model_sharding = True,
		n_feature = model_config.dim,
		n_head = model_config.n_heads,
		n_kv_head = model_config.n_kv_heads
    )
```
Thanks for all the great suggestions!

Best,
Yushun
